src/devices/sound/qsoundhle.cpp

```diff
@@ -180,70 +212,429 @@ WRITE8_MEMBER(qsound_hle_device::qsound_w)
+		register_map[i+0xba] = (uint16_t*)&m_voice[i].echo;
 	}
-	else if (address < 0x90)
+
+	// ADPCM registers
+	for(int i=0;i<3;i++) // ADPCM voices
 	{
-		ch = address & 0xf;
-		reg = 8;
+		// ADPCM sample rate is fixed to 8khz. (one channel is updated every third sample)
+		register_map[(i<<2)+0xca] = (uint16_t*)&m_adpcm[i].start_addr;
+		register_map[(i<<2)+0xcb] = (uint16_t*)&m_adpcm[i].end_addr;
+		register_map[(i<<2)+0xcc] = (uint16_t*)&m_adpcm[i].bank;
+		register_map[(i<<2)+0xcd] = (uint16_t*)&m_adpcm[i].volume;
+		register_map[i+0xd6] = (uint16_t*)&m_adpcm[i].flag; // non-zero to start ADPCM playback
+		register_map[i+0x90] = (uint16_t*)&m_voice_pan[16+i];
 	}
-	else if (address >= 0xba && address < 0xca)
+
+	// QSound registers
+	register_map[0x93] = (uint16_t*)&m_echo.feedback;
+	register_map[0xd9] = (uint16_t*)&m_echo.end_pos;
+	register_map[0xe2] = (uint16_t*)&m_delay_update; // non-zero to update delays
+	register_map[0xe3] = (uint16_t*)&m_next_state;
+	for(int i=0;i<2;i++)  // left, right
 	{
-		ch = address - 0xba;
-		reg = 9;
+		// Wet
+		register_map[(i<<1)+0xda] = (uint16_t*)&m_filter[i].table_pos;
+		register_map[(i<<1)+0xde] = (uint16_t*)&m_wet[i].delay;
+		register_map[(i<<1)+0xe4] = (uint16_t*)&m_wet[i].volume;
+		// Dry
+		register_map[(i<<1)+0xdb] = (uint16_t*)&m_alt_filter[i].table_pos;
+		register_map[(i<<1)+0xdf] = (uint16_t*)&m_dry[i].delay;
+		register_map[(i<<1)+0xe5] = (uint16_t*)&m_dry[i].volume;
+	}
+}
+
+int16_t qsound_hle_device::get_sample(uint16_t bank,uint16_t address)
+{
+	uint32_t rom_addr;
+	uint8_t sample_data;
+
+	bank &= 0x7FFF;
+	rom_addr = (bank << 16) | (address << 0);
+
+	sample_data = read_byte(rom_addr);
+
+	return (int16_t)(sample_data << 8);	// bit0-7 is tied to ground
+}
+
+/********************************************************************/
+
+// updates one DSP sample
+void qsound_hle_device::update_sample()
+{
+	switch(m_state)
+	{
+		default:
+		case STATE_INIT1:
+		case STATE_INIT2:
+			return state_init();
+		case STATE_REFRESH1:
+			return state_refresh_filter_1();
+		case STATE_REFRESH2:
+			return state_refresh_filter_2();
+		case STATE_NORMAL1:
+		case STATE_NORMAL2:
+			return state_normal_update();
+	}
+}
+
+// Initialization routine
+void qsound_hle_device::state_init()
+{
+	int mode = (m_state == STATE_INIT2) ? 1 : 0;
+
+	// we're busy for 4 samples, including the filter refresh.
+	if(m_state_counter >= 2)
+	{
+		m_state_counter = 0;
+		m_state = m_next_state;
+		return;
+	}
+	else if(m_state_counter == 1)
+	{
+		m_state_counter++;
+		return;
+	}
+
+	qsound_voice temp = {};		std::fill(std::begin(m_voice),std::end(m_voice),temp);
+	qsound_adpcm temp2 = {};	std::fill(std::begin(m_adpcm),std::end(m_adpcm),temp2);
+	qsound_fir temp3 = {};		std::fill(std::begin(m_filter),std::end(m_filter),temp3);
+								std::fill(std::begin(m_alt_filter),std::end(m_alt_filter),temp3);
+	qsound_delay temp4 = {};	std::fill(std::begin(m_wet),std::end(m_wet),temp4);
```

Comment:
You really need to provide member initialisation for the `qsound_voice` etc. structures for this to work properly, and you don't need the locals, as you can construct a temporary in the call, e.g. `std::fill(std::begin(m_voice), std::end(m_voice), qsound_voice());`